lib/galaxy/webapps/galaxy/services/workflows.py

```diff
@@ -105,6 +119,71 @@ def index(
             return workflows, total_matches
         return rval, total_matches
 
+    def invoke_workflow(
+        self,
+        trans,
+        workflow_id,
+        payload: InvokeWorkflowPayload,
+    ) -> Union[WorkflowInvocationResponse, List[WorkflowInvocationResponse]]:
+        # Get workflow + accessibility check.
+        by_stored_id = not payload.instance
+        stored_workflow = self._workflows_manager.get_stored_accessible_workflow(trans, workflow_id, by_stored_id)
+        workflow = stored_workflow.latest_workflow
+        run_configs = build_workflow_run_configs(trans, workflow, payload.model_dump(exclude_unset=True))
+        is_batch = payload.batch
+        if not is_batch and len(run_configs) != 1:
+            raise HTTPException(status_code=400, detail="Must specify 'batch' to use batch parameters.")
+
+        require_exact_tool_versions = payload.require_exact_tool_versions
+        tools = self._workflow_contents_manager.get_all_tools(workflow)
+        missing_tools = [
+            tool
+            for tool in tools
+            if not trans.app.toolbox.has_tool(
+                tool["tool_id"], tool_version=tool["tool_version"], exact=require_exact_tool_versions
+            )
+        ]
+        if missing_tools:
+            missing_tools_message = "Workflow was not invoked; the following required tools are not installed: "
+            if require_exact_tool_versions:
+                missing_tools_message += ", ".join(
+                    [f"{tool['tool_id']} (version {tool['tool_version']})" for tool in missing_tools]
+                )
+            else:
+                missing_tools_message += ", ".join([tool["tool_id"] for tool in missing_tools])
+            raise HTTPException(status_code=400, detail=missing_tools_message)
+        # if missing_tools:
+        #     missing_tools_message = "Workflow was not invoked; the following required tools are not installed: "
+        #     if require_exact_tool_versions:
+        #         missing_tools_message += ", ".join(
+        #             [f"{tool['tool_id']} (version {tool['tool_version']})" for tool in missing_tools]
+        #         )
+        #     else:
+        #         missing_tools_message += ", ".join([tool["tool_id"] for tool in missing_tools])
+        #     raise exceptions.MessageException(missing_tools_message)
```

Comment:
I replaced it with the code above:

```
        if missing_tools:
            missing_tools_message = "Workflow was not invoked; the following required tools are not installed: "
            if require_exact_tool_versions:
                missing_tools_message += ", ".join(
                    [f"{tool['tool_id']} (version {tool['tool_version']})" for tool in missing_tools]
                )
            else:
                missing_tools_message += ", ".join([tool["tool_id"] for tool in missing_tools])
            raise HTTPException(status_code=400, detail=missing_tools_message)
```

 and it is not used anymore. I was not sure if 400 is the appropriate response code, so I waited for review before removing it.

Reply:
Ahhh I see I got confused by the entire block instead of just the exception raising.

You should use MessageException instead of HTTPException, with that, you would also fix https://github.com/galaxyproject/galaxy/pull/17463#discussion_r1500924267